src/analyses/memLeak.ml

```diff
@@ -0,0 +1,87 @@
+(** An analysis for the detection of memory leaks ([memLeak]). *)
+
+open GoblintCil
+open Analyses
+open MessageCategory
+
+module ToppedVarInfoSet = SetDomain.ToppedSet(CilType.Varinfo)(struct let topname = "All Heap Variables" end)
+
+module Spec : Analyses.MCPSpec =
+struct
+  include Analyses.DefaultSpec
+
+  let name () = "memLeak"
+
+  module D = ToppedVarInfoSet
+  module C = Lattice.Unit
+
+  let context _ _ = ()
+
+  (* HELPER FUNCTIONS *)
+  let warn_for_multi_threaded ctx =
+    if not (ctx.ask (Queries.MustBeSingleThreaded { since_start = true })) then
+      M.warn ~category:(Behavior (Undefined MemoryLeak)) ~tags:[CWE 401] "Program isn't running in single-threaded mode. A memory leak might occur due to multi-threading"
+
+  (* TRANSFER FUNCTIONS *)
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    ctx.local
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    let state = ctx.local in
+    (* TODO: Is this too hacky of a solution? *)
```

Comment:
I think it is missing other ways in which a function may be left, such as `exit` and so on. Probably, handling of the `Abort` special function is needed here.

Reply:
Thanks for the tip! I previously tried matching the `Abort` function in the `special` transfer function and debug printing some stuff to see if anything happens upon termination via `exit`, but I didn't get anything meaningful there and so I didn't even consider it anymore afterwards.

I'll try `Abort` in the `return` transfer function

Reply:
No, it should in fact be handled by the `special` transfer function. Strange that it didn't show up!

Reply:
Forget about the previous comment. I got confused. I'll add the checks for `Abort` in `special` as I had tried earlier.

Reply:
I'll also add a simple check for `Assert` in case assert's expression can be evaluated to something which is 0, i.e., false.

Reply:
> Should we consider supporting them as well?

Yes, that would be a good idea! I'd propose to do it in a separate PR.

Reply:
> I'd propose to do it in a separate PR

Opened #1135 for `quick_exit` support.